adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1718,38 +1799,73 @@ private boolean stopCalled() {
         }
 
         /**
-         * Closes a connection to a client.
+         * Actively closes the connection to the device.
+         *
+         * @param reason The reason for closing the connection.
+         * @param sendDisconnectedEvent {@code true} if events shall be sent concerning the disconnect.
+         * @throws NullPointerException if reason is {@code null}.
+         */
+        protected final void close(final String reason, final boolean sendDisconnectedEvent) {
+            Objects.requireNonNull(reason);
+            endpoint.closeHandler(v -> {});
+            onCloseInternal("close device connection", reason, sendDisconnectedEvent);
+        }
+
+        /**
+         * To be called by the MqttEndpoint closeHandler.
          */
         protected final void onClose() {
-            final String operationName = stopCalled() ? "CLOSE on server shutdown" : "CLOSE";
-            final Span span = newSpan(operationName);
+            if (stopCalled()) {
+                onCloseInternal("close device connection (server shutdown)", "server shutdown", true);
+            } else {
+                onCloseInternal("CLOSE", null, true);
+            }
+        }
+
+        private void onCloseInternal(final String spanOperationName, final String reason, final boolean sendDisconnectedEvent) {
+            final Span span = newSpan(spanOperationName);
+            AbstractVertxBasedMqttProtocolAdapter.this.onBeforeEndpointClose(this);
             AbstractVertxBasedMqttProtocolAdapter.this.onClose(endpoint);
-            final CompositeFuture removalDoneFuture = removeAllCommandSubscriptions(span);
-            sendDisconnectedEvent(endpoint.clientIdentifier(), authenticatedDevice, span.context());
+            final CompositeFuture removalDoneFuture = removeAllCommandSubscriptions(span, sendDisconnectedEvent);
+            if (sendDisconnectedEvent) {
+                sendDisconnectedEvent(endpoint.clientIdentifier(), authenticatedDevice, span.context());
+            }
             if (authenticatedDevice == null) {
-                log.debug("connection to anonymous device [clientId: {}] closed", endpoint.clientIdentifier());
                 metrics.decrementUnauthenticatedConnections();
             } else {
-                log.debug("connection to device [tenant-id: {}, device-id: {}] closed",
-                        authenticatedDevice.getTenantId(), authenticatedDevice.getDeviceId());
                 metrics.decrementConnections(authenticatedDevice.getTenantId());
             }
+            final String reasonSuffix = reason != null ? ("; reason: " + reason) : "";
             if (endpoint.isConnected()) {
-                log.debug("closing connection with client [client ID: {}]", endpoint.clientIdentifier());
+                span.log("close device connection" + reasonSuffix);
```

Comment:
instead of requiring a user to *parse* the string you are logging, we could log a Map to the span containing keys Fields.MESSAGE and (optionally) `reason` ...

Reply:
I've adapted the code accordingly (using `Fields.EVENT`, fits better FMPOV).